### PR TITLE
Drop support for empty application binding

### LINKED ISSUE #897: Drop support for no application binding
Reference EPIC #892 
Use Case: 
Support declaring no application in binding: generate intermediate binding secret only

Reasons:

Dropping it, as this would require the ServiceBinding resource mutable.
Dropping it, this could be handle with https://github.com/k8s-service-bindings/spec#binding-secret-generation-strategies
 
Requirements:

Will need to provide documentation about a different path to achieve the “same use case”

### PR DESCRIPTION
**Pull request by baijum**
Fixes #897

### CODE REVIEW

File: test/acceptance/features/bindAppToService.feature [hunk 1]
```diff
@@ -330,10 +330,8 @@ Feature: Bind an application to a service
             """
         Then jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding "binding-request-empty-app" should be changed to "True"
         And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-empty-app" should be changed to "False"
-        And jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding "binding-request-empty-app" should be changed to "EmptyApplication"
-        And jq ".status.conditions[] | select(.type=="Ready").status" of Service Binding "binding-request-empty-app" should be changed to "True"
-        And Secret contains "BACKEND_HOST" key with value "example.common"
-        And Secret contains "BACKEND_USERNAME" key with value "foo"
+        And jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding "binding-request-empty-app" should be changed to "ApplicationNotFound"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:333]: if application selector is not specified, we should not come this far at all - at service binding creation we should get the error.

File: test/acceptance/features/bindAppToService.feature [hunk 2]
```diff
@@ -679,8 +677,8 @@ Feature: Bind an application to a service
             """
         * jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding "binding-request-remove-service" should be changed to "True"
         * jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-remove-service" should be changed to "False"
-        * jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding "binding-request-remove-service" should be changed to "EmptyApplication"
-        * jq ".status.conditions[] | select(.type=="Ready").status" of Service Binding "binding-request-remove-service" should be changed to "True"
+        * jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding "binding-request-remove-service" should be changed to "ApplicationNotFound"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:680]: see above. This scenario is not possible at all anymore.

**Review by pedjak — CHANGES_REQUESTED** [review]
since `Context.Applications()` returns now error if with the given application selector no applications are found, we should add an acceptance test demonstrating this.

File: pkg/reconcile/pipeline/builder/pipeline_integration_test.go [hunk 3]
```diff
@@ -152,7 +153,7 @@ var _ = Describe("Default Pipeline", func() {
 		updatedApp := &appsv1.Deployment{}
 		err = runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, updatedApp)
 		Expect(err).NotTo(HaveOccurred())
-		Expect(updatedApp.Spec.Template.Spec.Containers[0].EnvFrom[0].SecretRef.Name).To(Equal(updatedSB.Status.Secret))
+		//Expect(updatedApp.Spec.Template.Spec.Containers[0].EnvFrom[0].SecretRef.Name).To(Equal(updatedSB.Status.Secret))
```
Comment (pedjak) [inline pkg/reconcile/pipeline/builder/pipeline_integration_test.go:156]: if not needed, please remove.

File: pkg/reconcile/pipeline/context/application.go [hunk 4]
```diff
@@ -1,10 +1,11 @@
 package context
 
 import (
+	"reflect"
+
 	"github.com/redhat-developer/service-binding-operator/api/v1alpha1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"reflect"
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/application.go:7]: no-op change: only import order changed - please revert.

File: pkg/reconcile/pipeline/context/impl.go [hunk 5]
```diff
@@ -159,9 +159,18 @@ func (i *impl) Applications() ([]pipeline.Application, error) {
...
+	if len(result) == 0 {
+		return nil, emptyApplicationsErr{}
+	}
 	return result, nil
 }
 
+type emptyApplicationsErr struct{}
+
+func (e emptyApplicationsErr) Error() string {
+	return "Empty applications"
```
[the remark this answers is not in the document]
Comment (baijum) [reply pkg/reconcile/pipeline/context/impl.go]: Done. 2859eea3a0dca45b48a47fa3d59c1b34b6ff9cc7

File: api/v1alpha1/servicebinding_types.go [hunk 6]
```diff
@@ -86,7 +86,7 @@ type ServiceBindingSpec struct {
 	// Application is used to identify the application connecting to the
 	// backing service operator.
 	// +optional
-	Application *Application `json:"application,omitempty"`
+	Application *Application `json:"application"`
```
[the remark this answers is not in the document]
Comment (baijum) [reply api/v1alpha1/servicebinding_types.go]: Done.

File: test/acceptance/features/bindAppToService.feature (hunk 1 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature]: Fixed.

File: test/acceptance/features/bindAppToService.feature (hunk 2 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature]: Fixed.

File: test/acceptance/features/bindAppToService.feature [hunk 7]
```diff
@@ -328,16 +328,13 @@ Feature: Bind an application to a service
...
-        And jq ".status.conditions[] | select(.type=="Ready").status" of Service Binding "binding-request-empty-app" should be changed to "True"
-        And Secret contains "BACKEND_HOST" key with value "example.common"
-        And Secret contains "BACKEND_USERNAME" key with value "foo"
+        Then Error message is thrown
+        And Service Binding "binding-request-empty-app" is not persistent in the cluster
 
     @olm
     Scenario: Backend Service new spec status update gets propagated to the binding secret
         Given OLM Operator "backend-new-spec" is running
+        * Generic test application "gen-app-a-s-c" is running
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:337]: this scenario should be removed because we do not support just creating binding secrets if application is not specified.

File: test/acceptance/features/bindAppToService.feature [hunk 8]
```diff
@@ -512,10 +513,14 @@ Feature: Bind an application to a service
                     version: v1
                     kind: Backserv
                     name: demo-backserv-cr-2
+                application:
+                    name: gen-app-a-s-d
+                    group: apps
+                    version: v1
+                    resource: deployments
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:520]: this scenario should be now either renamed or removed, because adding the application references changes the scenario semantic.

File: test/acceptance/features/bindAppToService.feature [hunk 9]
```diff
@@ -661,8 +666,9 @@ Feature: Bind an application to a service
             apiVersion: "stable.example.com/v1"
             kind: Backend
             metadata:
-                name: demo-backserv-cr-2
+                name: demo-backserv-cr-3
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:669]: why this change?

File: test/acceptance/features/bindAppToService.feature [hunk 10]
```diff
@@ -689,6 +697,11 @@ Feature: Bind an application to a service
                 name: binding-request-remove-service
             spec:
                 services:
+                application:
+                    name: gen-app-a-s-e
+                    group: apps
+                    version: v1
+                    resource: deployments
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:704]: why this change? even without the application reference, the service binding was invalid.

File: test/acceptance/features/bindAppToServiceAnnotations.feature [hunk 11]
```diff
@@ -58,13 +59,17 @@ Feature: Bind an application to a service using annotations
...
+                    version: v1
+                    resource: deployments
             """
-        Then jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "True"
-        And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "False"
-        And jq ".status.conditions[] | select(.type=="InjectionReady").reason" of Service Binding "binding-request-backend-a" should be changed to "EmptyApplication"
+
+        Then Service Binding "binding-request-backend-a" is ready
+        And jq ".status.conditions[] | select(.type=="CollectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "True"
+        And jq ".status.conditions[] | select(.type=="InjectionReady").status" of Service Binding "binding-request-backend-a" should be changed to "True"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToServiceAnnotations.feature:71]: redundant: "Service Binding "binding-request-backend-a" is ready" covers these checks.

File: test/acceptance/features/bindAppToServiceAnnotations.feature [hunk 12]
```diff
@@ -58,13 +59,17 @@ Feature: Bind an application to a service using annotations
...
         And jq ".status.conditions[] | select(.type=="Ready").status" of Service Binding "binding-request-backend-a" should be changed to "True"
-        And Secret contains "BACKEND_READY" key with value "true"
-        And Secret contains "BACKEND_HOST" key with value "example.common"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToServiceAnnotations.feature:67]: we should check if the application sees the appropriate bindings. Also scenario should be renamed to reflect the change in the usecase.

File: test/acceptance/features/steps/openshift.py [hunk 13]
```diff
@@ -185,6 +185,7 @@ def apply(self, yaml, namespace=None):
         return output
 
     def apply_invalid(self, yaml, namespace=None):
+        print("AAAAAAAAAAAA", yaml)
```
Comment (pedjak) [inline test/acceptance/features/steps/openshift.py:188]: this should be removed, right?

**Review by pedjak — CHANGES_REQUESTED** [review]
on top of this, please check why `api/v1alpha1/zz_generated.deepcopy.go` is removed in this PR. IMHO, that should not be the case.

**Comment by baijum** [comment]
> on top of this, please check why `api/v1alpha1/zz_generated.deepcopy.go` is removed in this PR. IMHO, that should not be the case.

That file is not removed. Only a few lines removed (autogenerated).

File: pkg/reconcile/pipeline/context/application.go (hunk 4 above)
Comment (baijum) [reply pkg/reconcile/pipeline/context/application.go:7]: I have integrated [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) with my editor. So, this change happens when I save the file. Can I send a separate PR with these changes?

File: test/acceptance/features/bindAppToService.feature (hunk 7 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:337]: This scenario is different, please read the scenario once again. The empty application was used earlier for convenience.

File: test/acceptance/features/bindAppToService.feature (hunk 8 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:520]: This is a valid scenario. The empty application was used earlier for convenience.

File: test/acceptance/features/bindAppToService.feature (hunk 9 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:669]: The `demo-backserv-cr-2` resource name is already used. And it was making the test more fragile. A unique name is always better.

File: test/acceptance/features/bindAppToService.feature (hunk 10 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:704]: This a valid negative scenario. See the title: `Removing service from services field from existing serivce binding is not allowed`

File: pkg/reconcile/pipeline/context/application.go (hunk 4 above)
Comment (pedjak) [reply pkg/reconcile/pipeline/context/application.go:7]: but this is the only change on this file, not sure why you needed to save it? Please revert it. If we want to fix import ordering globally, that should be done in a separate PR, but we should agree about the common rules that will be used by majority of IDEs/editors we use.

File: test/acceptance/features/steps/openshift.py (hunk 13 above)
Comment (baijum) [reply test/acceptance/features/steps/openshift.py]: Removed. c46fcb9aa05b6bd0887a6a8dab12615d9fc47c1f

File: pkg/reconcile/pipeline/builder/pipeline_integration_test.go (hunk 3 above)
Comment (baijum) [reply pkg/reconcile/pipeline/builder/pipeline_integration_test.go]: Reverted. 1765cebdf501d466932aaa6708621ff5d3a3e087
(I commented it earlier as it was failing)

File: test/acceptance/features/bindAppToServiceAnnotations.feature (hunk 12 above)
Comment (baijum) [reply test/acceptance/features/bindAppToServiceAnnotations.feature:67]: Done. 5848c6ad9efcae4fd19a89610c4b87091efb3376

File: test/acceptance/features/bindAppToServiceAnnotations.feature (hunk 11 above)
Comment (baijum) [reply test/acceptance/features/bindAppToServiceAnnotations.feature]: Done. 5848c6ad9efcae4fd19a89610c4b87091efb3376

File: pkg/reconcile/pipeline/context/application.go (hunk 4 above)
Comment (baijum) [reply pkg/reconcile/pipeline/context/application.go:7]: I send separate PR: https://github.com/redhat-developer/service-binding-operator/pull/981

Comment (baijum) [reply pkg/reconcile/pipeline/context/application.go]: Reverted here: 6ee0d5dedefe8d356fcee88eebd9fea701325b79

File: pkg/reconcile/pipeline/builder/pipeline_integration_test.go [hunk 14]
```diff
@@ -16,7 +18,6 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/dynamic/fake"
-	"reflect"
```
Comment (baijum) [inline pkg/reconcile/pipeline/builder/pipeline_integration_test.go:19]: @pedjak Let me know if I should revert this change.

File: test/acceptance/features/bindAppToService.feature (hunk 7 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:337]: The scenario description is "Backend Service new spec status update gets propagated to the binding secret" and IMHO it is obsolete now because now we have introduced the application ref in the whole game. Hence, the scenario description should be renamed to reflect that, e,g,:

"Bind service to application using binding definition available in x-descriptors"

And to assert that, we should drop lines 372-374 and check instead what application sees.

File: test/acceptance/features/bindAppToService.feature (hunk 8 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:520]: Similar to above - by introducing now application ref, the usecase is not the one described by the scenario anymore:

"Creating binding secret from the definitions managed in OLM operator descriptors"

We do not just create secret anymore, we inject binding data into the application. That should be reflected into the scenario description and the appropriate assertions.

**Review by pedjak — CHANGES_REQUESTED** [review]

File: pkg/reconcile/pipeline/context/impl.go [hunk 15]
```diff
@@ -122,49 +122,56 @@ func (i *impl) Services() ([]pipeline.Service, error) {
...
+			}
+
+			emptyResult := make([]pipeline.Application, 0)
+			objList, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).List(context.Background(), opts)
+			if err != nil {
+				return emptyResult, err
+			}
+
+			if len(objList.Items) == 0 {
+				return emptyResult, nil
```
[the remark this answers is not in the document]
Comment (baijum) [reply pkg/reconcile/pipeline/context/impl.go]: Done. c0cd26e9195b909d7679dab6f8351641d83f5880

File: pkg/reconcile/pipeline/context/impl.go [hunk 16]
```diff
@@ -122,49 +122,56 @@ func (i *impl) Services() ([]pipeline.Service, error) {
...
-			i.applications = make([]*application, 0)
 		}
 	}
+
 	result := make([]pipeline.Application, len(i.applications))
 	for l, a := range i.applications {
 		result[l] = a
 	}
+	if len(result) == 0 {
+		return nil, emptyApplicationsErr{}
```
[the remark this answers is not in the document]
Comment (baijum) [reply pkg/reconcile/pipeline/context/impl.go]: Done. c0cd26e9195b909d7679dab6f8351641d83f5880

File: test/acceptance/features/bindAppToService.feature (hunk 7 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:337]: Done. 29d348dd24be04f7ef01892df9186f888247adf1

File: test/acceptance/features/bindAppToService.feature (hunk 8 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:520]: Done. 29d348dd24be04f7ef01892df9186f888247adf1

File: pkg/reconcile/pipeline/context/impl.go [hunk 17]
```diff
@@ -120,51 +120,61 @@ func (i *impl) Services() ([]pipeline.Service, error) {
...
+		}
+		if i.serviceBinding.Spec.Application.LabelSelector != nil && i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+			matchLabels := i.serviceBinding.Spec.Application.LabelSelector.MatchLabels
+			opts := metav1.ListOptions{
+				LabelSelector: labels.Set(matchLabels).String(),
+			}
+
+			objList, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).List(context.Background(), opts)
+			if err != nil {
+				return nil, emptyApplicationsErr{err}
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/impl.go:147]: we do not know if application list is empty or not, we should just pass through the obtained error to the caller:

```
return nil, err
```

File: pkg/reconcile/pipeline/context/impl.go [hunk 18]
```diff
@@ -120,51 +120,61 @@ func (i *impl) Services() ([]pipeline.Service, error) {
 	return services, nil
 }
 
+// Application return a list of applications.
+// And if no application found, return an error
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/impl.go:124]: please update  the interface doc in`pkg/reconcile/pipeline/api.go`

File: test/acceptance/features/bindAppToService.feature [hunk 19]
```diff
@@ -416,7 +416,7 @@ Feature: Bind an application to a service
         And The application env var "HOST_ADDR" has value "127.0.0.1:8080"
 
     @olm
-    Scenario: Creating binding secret from the definitions managed in OLM operator descriptors
+    Scenario: Bind service to application using binding definition available in CSV x-descriptors
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:419]: this scenario and the one in line 335 talk about the same thing? what is the difference?

**Review by pedjak — APPROVED** [review]
over look good - a few small open things to be solved and CI jobs needs to pass.

File: pkg/reconcile/pipeline/context/impl.go (hunk 17 above)
Comment (baijum) [reply pkg/reconcile/pipeline/context/impl.go]: Done. 34cd1e9bf29f243191f69e5352b0440b288b3927

File: pkg/reconcile/pipeline/context/impl.go (hunk 18 above)
Comment (baijum) [reply pkg/reconcile/pipeline/context/impl.go:124]: Done. 34cd1e9bf29f243191f69e5352b0440b288b3927

File: test/acceptance/features/bindAppToService.feature (hunk 19 above)
Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:419]: The use of x-descriptors is different.  Earlier one use:
`service.binding:ports:elementType=sliceOfMaps,sourceKey=protocol,sourceValue=port`

File: pkg/reconcile/pipeline/context/impl_test.go [hunk 20]
```diff
@@ -233,6 +222,7 @@ var _ = Describe("Context", func() {
 
 			_, err := ctx.Applications()
 			Expect(err).To(HaveOccurred())
+			fmt.Println(err.Error())
```
Comment (pedjak) [inline pkg/reconcile/pipeline/context/impl_test.go:225]: do we really need this line?

Comment (baijum) [reply pkg/reconcile/pipeline/context/impl_test.go]: removed 7c194b11d0a69dc090b73eed1edae3707e476d68

File: test/acceptance/features/bindAppToService.feature (hunk 19 above)
Comment (pedjak) [reply test/acceptance/features/bindAppToService.feature:419]: but both of them define bindings via x-descriptors and they descriptions do not tell any other details:

* " Bind service to application using binding definition available in CSV x-descriptors "
* "Bind service to application using binding definition available in x-descriptors"

The descriptions are almost identical. Would it make sense to either remove one of the scenarios, or collapse them into single one?

Comment (baijum) [reply test/acceptance/features/bindAppToService.feature:419]: Let's remove the current one - it looks like a simple case.

Comment (baijum) [reply test/acceptance/features/bindAppToService.feature]: removed.

**Comment by pedjak** [comment]
/lgtm

letting @isutton to double check and approve it finally

File: pkg/reconcile/pipeline/context/impl_test.go [hunk 21]
```diff
@@ -45,15 +44,6 @@ var _ = Describe("Context", func() {
...
-			sb := v1alpha1.ServiceBinding{
-				Spec: v1alpha1.ServiceBindingSpec{},
-			}
-			ctx := &impl{serviceBinding: &sb}
-			applications, err := ctx.Applications()
-			Expect(err).NotTo(HaveOccurred())
-			Expect(applications).To(BeEmpty())
-		})
 
 		DescribeTable("should return slice of size 1 if application is specified", func(bindingPath *v1alpha1.BindingPath, expectedContainerPath string) {
```
Comment (isutton) [inline pkg/reconcile/pipeline/context/impl_test.go:48]: This description should be changed, now that application is always specified.

File: test/acceptance/features/bindAppToService.feature [hunk 22]
```diff
@@ -415,113 +415,6 @@ Feature: Bind an application to a service
         Then Service Binding "service-binding-c-e" is ready
         And The application env var "HOST_ADDR" has value "127.0.0.1:8080"
 
-    @olm
-    Scenario: Creating binding secret from the definitions managed in OLM operator descriptors
```
Comment (isutton) [inline test/acceptance/features/bindAppToService.feature:419]: Why is this scenario being removed? Isn't the case of filling the application in the ServiceBinding below?

It seems this scenario didn't provide an application because its presence didn't influence on the results being observed, so removing this might not be what we want, unless it has been completely deprecated.

**Review by isutton — APPROVED** [review]
There is a question regarding removing an acceptance test, but in general LGTM.
